webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+func ProcessFirstTime(rdb *redis.Client, mu *sync.Mutex) {
+	ctx := context.Background()
+	for {
+		results, err := rdb.BLPop(ctx, 0, firstTimeQueue).Result()
+		if err != nil {
+			logger.Error(fmt.Sprintf("First-time pop error: %v", err))
+			time.Sleep(time.Second)
+			continue
+		}
+
+		var msg Message
+		rawMessage := []byte(results[1])
+		if err := json.Unmarshal(rawMessage, &msg); err != nil {
+			logger.Error(fmt.Sprintf("First-time unmarshal error: %v, JSON: %s", err, rawMessage))
+			continue
+		}
+
+		webhookID, err := ExtractWebhookID(msg.Payload.URL)
+		if err != nil {
+			logger.Error(fmt.Sprintf("Could not parse webhook ID from: %s:%s", msg.Payload.URL, err))
+			continue
+		}
+
+		// If we've already seen this webhookID and know the bucket, insert it
+		// into the appropriate queue instead of sending here
+		bucket, err := rdb.HGet(ctx, bucketsHash, webhookID).Result()
+		if err != redis.Nil {
+			queueKey := bucketQueuePrefix + bucket
+			var msgBytes []byte
+			msgBytes, _ = json.Marshal(msg)
+			rdb.RPush(ctx, queueKey, msgBytes)
+			rdb.LTrim(ctx, queueKey, 0, maxQueueSize-1)
+			continue
+		}
+
+		// Our retry logic here does block the first time queue but this should be
+		// fairly rare; after any successful attempt messages will be routed to the
+		// appropriate queue; and there aren't that many unique places CRCON dispatches
+		// webhooks from
+		rateLimit := &RateLimitState{}
+		for attempts := 0; attempts < maxMsgReattempts && (!msg.Discardable); attempts++ {
+			bucket, err = SendWebhook(rdb, &msg, rateLimit)
+			if err == nil {
+				rdb.HSet(ctx, bucketsHash, webhookID, bucket)
+				// Start worker if new bucket
+				if _, exists := GetWorker(bucket, mu); !exists {
+					// Make a new worker but use the response we got from Discord
+					// for the first time the message was sent
+					worker := NewBucketWorker(bucket, rdb)
+					worker.rateLimit.RateLimited = rateLimit.RateLimited
+					worker.rateLimit.Remaining = rateLimit.Remaining
+					worker.rateLimit.ResetAfter = rateLimit.ResetAfter
+					worker.rateLimit.ResetTime = rateLimit.ResetTime
+					SetWorker(bucket, worker, mu)
+					go worker.ProcessQueue(rdb)
+				}
+			} else {
+				logger.Error(fmt.Sprintf("First-time send error: %v", err))
+				if rateLimit.RateLimited {
+					wait := GetWorkerRateLimitSleepTime(rateLimit)
+					time.Sleep(wait)
+				} else {
+					// We had some sort of (non rate limiting error) trying to send this message
+					// back off before attempting again
+					time.Sleep(time.Second * 2 * time.Duration(attempts+1))
+				}
+			}
+		}
+		queueKey := bucketQueuePrefix + bucket
+		rdb.LTrim(ctx, queueKey, 0, maxQueueSize-1)
+	}
+}
+
+func ProcessTransient(rdb *redis.Client, msg Message, mu *sync.Mutex) {
+	webhookID, err := ExtractWebhookID(msg.Payload.URL)
+	if err != nil {
+		logger.Error(fmt.Sprintf("could not parse webhook ID from: %s:%s", msg.Payload.URL, err))
+		return
+	}
+
+	ctx := context.Background()
```

Comment:
I see a lot of `context.Background` calls flying around here. The better approach would be to have a single context passed around for the calls to be made. Meaning: In the entrypoint method, use `context.Background` (or any other way to construct a context for that matter) and pass that as an argument to the methods that require a context. Usually this is the first argument to a method.